src/resolvers/libsodium.rs

```diff
@@ -0,0 +1,280 @@
+extern crate sodiumoxide;
+
+use byteorder::{ByteOrder, LittleEndian};
+
+use params::{CipherChoice, DHChoice, HashChoice};
+use types::{Cipher, Dh, Hash, Random};
+use CryptoResolver;
+
+use self::sodiumoxide::crypto::aead::chacha20poly1305_ietf as sodium_chacha20poly1305;
+use self::sodiumoxide::crypto::hash::sha256 as sodium_sha256;
+use self::sodiumoxide::crypto::scalarmult::curve25519 as sodium_curve25519;
+
+#[derive(Default)]
+pub struct SodiumResolver;
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<Random + Send>> {
+        None
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<Dh + Send>> {
+        match *choice {
+            DHChoice::Curve25519 => Some(Box::new(SodiumDh25519::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<Hash + Send>> {
+        match *choice {
+            HashChoice::SHA256 => Some(Box::new(SodiumSha256::default())),
+            _ => None,
+        }
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<Cipher + Send>> {
+        match *choice {
+            CipherChoice::ChaChaPoly => Some(Box::new(SodiumChaChaPoly::default())),
+            _ => None,
+        }
+    }
+}
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
+}
+
+impl SodiumDh25519 {
+    fn convert_to_private_key(key: &mut [u8; 32]) {
+        key[0] &= 248;
+        key[31] &= 127;
+        key[31] |= 64;
+    }
+}
+
+impl Default for SodiumDh25519 {
+    fn default() -> SodiumDh25519 {
+        SodiumDh25519 {
+            privkey: sodium_curve25519::Scalar([0; 32]),
+            pubkey: sodium_curve25519::GroupElement([0; 32]),
+        }
+    }
+}
+
+impl Dh for SodiumDh25519 {
+    fn name(&self) -> &'static str {
+        "25519"
+    }
+
+    fn pub_len(&self) -> usize {
+        32
+    }
+
+    fn priv_len(&self) -> usize {
+        32
+    }
+
+    fn set(&mut self, privkey: &[u8]) {
+        self.privkey = sodium_curve25519::Scalar::from_slice(privkey)
+            .expect("Can't construct private key for Dh25519");
```

Comment:
when exactly would this cause a panic?

Reply:
When slice of incorrect length is passed, for example.

Reply:
Cool thanks. Seems reasonable.